### ISSUE TITLE
[Feature Request] We need a method to Clear Token Cache for Confidential Client

### ISSUE THREAD

**Issue by msJinLei** [AzureAD/microsoft-authentication-library-for-dotnet#4854]
### MSAL client type

Confidential

### Problem statement

The customers sometimes  want to actively clear token cache for security concern. We have receive several asks from the customers. However,  there is currently no way for us to clear app token cache.
MSAL.net does provide the interface to remove app token by account. But don't provide the interface to list all the accounts from confidential client.
```csharp
        /// <summary>
        /// Get the <see cref="IAccount"/> by its identifier among the accounts available in the token cache and of the same
        /// environment (authority host) as <see cref="Authority"/>.
        /// </summary>
        /// <param name="identifier">Account identifier. The value of the identifier will probably have been stored value from the
        /// value of the <see cref="AccountId.Identifier"/> property of <see cref="AccountId"/>.
        /// You typically get the account ID from an <see cref="IAccount"/> by using the <see cref="IAccount.HomeAccountId"/> property.</param>
        Task<IAccount> GetAccountAsync(string identifier);
```
### 

### Proposed solution

- Provide an interface to clear app token cache and user token cache

### Alternatives

- or Provide an interface to just clear app token cache
- or Provide an interface to list all the accounts created from confidential client.

**Comment by isra-fel** [comment]
The two options to clear (app) token cache are both obsolete.

```csharp
        private async Task ClearCacheAsync()
        {
            StorageCreationProperties storageProperties;
            (await MsalCacheHelper.CreateAsync(storageProperties)).Clear(); // obsolete

            IConfidentialClientApplication app;
            foreach (var account in await app.GetAccountsAsync()) // also obsolete
            {
                await app.RemoveAsync(account);
            }
        }
```

**Comment by bgavrilMS** [comment]
A similar ask was done on MSAL Python - https://github.com/AzureAD/microsoft-authentication-library-for-python

**Comment by bgavrilMS** [comment]
CC @localden  and @jmprieur for the feature request from Az.PS team. Similar request was made on Az.CLI team.

**Comment by rayluo** [comment]
Indeed, a similar ask was done by `remove_tokens_for_client()` on [MSAL Python PR 666](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/666/files).

> ### Proposed solution
> * Provide an interface to clear app token cache and user token cache
> 
> ### Alternatives
> * or Provide an interface to just clear app token cache
> * or Provide an interface to list all the accounts created from confidential client.

Note that the `remove_tokens_for_client()` removes *app tokens* (a.k.a. "service-to-service token") only. But there is already a [`GetAccountsAsync()` that can return a list of all user accounts](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.62.0/src/client/Microsoft.Identity.Client/ClientApplicationBase.cs#L57-L63).

**Comment by austindonnelly** [comment]
Example from Azure Storage Explorer: when I've just finished a PIM activation of group membership granting me additional access, I want to refresh my token so the new rights are available.  Today, that's not possible, so the work around is to remove my user account, and re-add it.  This is quite painful from a UI point of view: many clicks involved.  The simple UI option ("Refresh token") is not possible, because MSAL doesn't have a way to invalidate the token cache.
Please add this, so Azure Storage Explorer can address their issue (linked above by cveId).
Thank you!

**Comment by isra-fel** [comment]
> [@isra-fel](https://github.com/isra-fel) - can you please say a few more words about the scenario? Why do customers want to reomve app tokens?

Sorry for the late response @bgavrilMS the idea behind this feature request is simply that customers are worried about token being cached and persistent after whatever job is done, and they want it cleared to eliminate any risk of exposure.
The most common scenario is probably a shared automation agent (a VM for example) where different apps sign in to.

Appreciate if you could prioritize this

**Comment by austindonnelly** [comment]
In my case, it's because I've just done a PIM activation, and the cached token doesn't have the extra group I am now a member of.  Getting rid of the cached token means I can re-authenticate, and pickup the enhanced rights I now have.

**Comment by rayluo** [comment]
> Example from Azure Storage Explorer: when I've just finished a PIM activation of group membership granting me additional access, I want to refresh my token so the new rights are available. Today, that's not possible, so the work around is to remove my user account, and re-add it. This is quite painful from a UI point of view: many clicks involved. The simple UI option ("Refresh token") is not possible, because MSAL doesn't have a way to invalidate the token cache. Please add this, so Azure Storage Explorer can address their issue (linked above by cveId). Thank you!

> In my case, it's because I've just done a PIM activation, and the cached token doesn't have the extra group I am now a member of. Getting rid of the cached token means I can re-authenticate, and pickup the enhanced rights I now have.

I would suggest @austindonnelly to create a different feature request for your case, if needed. Because in your case,

1. your app is not using "App Token Cache", it is using User Token Cache instead; the workaround is [`RemoveAsync(...)`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identity.client.clientapplicationbase.removeasync?view=msal-dotnet-latest) which you can use. However,
2. "Clear (entire?!) user token cache" would likely be an overkill because it throws away other valid tokens and results in an extra sign-in. Ideally, to skip one specific token which is known to be faulty. Currently, you shall try [`WithForceRefresh()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identity.client.acquiretokensilentparameterbuilder.withforcerefresh?view=msal-dotnet-latest#microsoft-identity-client-acquiretokensilentparameterbuilder-withforcerefresh(system-boolean)). (For what it's worth, we also have more precise way to [refresh one faulty token for Confidential Client](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.71.1/CHANGELOG.md?plain=1#L26-L29) (available in a [different namespace](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.71.1/src/client/Microsoft.Identity.Client/Extensibility/RP/AcquireTokenForClientParameterBuilderForResourceProviders.cs#L8)), but again, it does not apply if you are using Public Client.)
